Question_Categorization/clustering_and_tagging_classes.py

```diff
@@ -0,0 +1,298 @@
+    def get_top_n_keywords(self, n, questions):
+        '''
+            Finds the keywords in the set of questions that can be used as tags. 
+
+            Input:
+                n: Number of keywords to fetch. 
+                questions: List of strings.
+            
+            Output: 
+                top_n: List of strings where each string can be treated as a tag.
+        '''
+        try:
+            #Get the TF-IDF matrix for the questions.
+            preprocessed_questions = General().preprocess(questions, tokenize=False)
+            vectorizer = TfidfVectorizer(ngram_range=(1,2))
+            doc_term_matrix = vectorizer.fit_transform(preprocessed_questions)
+            all_terms = vectorizer.get_feature_names()
+            doc_term_array = doc_term_matrix.toarray()
+            
+            term_scores = [0 for i in range(0,len(all_terms))]
+
+            # Fetch the 40 highest scoring terms by adding the scores received by the term for all questions in the tf-idf matrix.
+            for i in range(0, len(questions)):
+                doc = doc_term_array[i]
+                for term_index in range(0, len(doc)):
+                    term_scores[term_index] += doc[term_index]
+            
+            term_scores = np.array(term_scores)
+            top_terms = (-term_scores).argsort()[:40]
+
+            #Find the frequency of the top 40 terms in the English language
+            freqs = []
+            for term in top_terms:
+                try:
+                    freqs.append(word_frequency(all_terms[term],"en"))
+                except:
+                    pass
+
+            #Fetch the top n rarest terms in the English language out of the set of 40 terms.
+            final_n = n
+            freqs = np.array(freqs)
+            top = (freqs).argsort()[:final_n]
+            roots = []
+            for term in top:
+                roots.append(all_terms[top_terms[term]])
+            
+            top_n = []
+            for term in top_terms:
+                top_n.append(all_terms[term])
+            
+            return top_n
+        
+        except Exception as e:
+            print('Error: ',e)
+            raise e
+
+    def complement_naive_bayes(self, roots, questions):
+        '''
+            Maps tags to question embeddings based on the Complement Naive Bayes algorithm. 
+
+            Input: 
+                questions: List of strings.
+                roots: List of strings, where each string is a tag.
+            
+            Output: 
+                root_to_ques_dict: A dictionary with tags as keys and a lists of questions as values. It indicates which questions can be mapped to which tags.
+        '''
+        try:
+            embeddings, embed = General().universal_sentence_encoder(questions, True)
+            #Add 1 to each value because CNB cannot work with negative values, and the embedding values never go below -1.
+            adjusted_embeddings = embeddings + 1
```

Comment:
One other thought - if we want to keeping using universal_sentence_encoder for the embeddings, might it make sense to use inner products between the keyword embeddings and document embeddings to determine which ones fit best instead of Complement Naive Bayes?